templates/commands/render_action_regexreplace.go

```diff
@@ -17,10 +17,144 @@ package commands
 import (
 	"context"
 	"fmt"
+	"regexp"
+	"strconv"
 
 	"github.com/abcxyz/abc/templates/model"
 )
 
-func actionRegexReplace(ctx context.Context, p *model.RegexReplace, sp *stepParams) error {
-	return fmt.Errorf("not implemented")
+// The regex_replace action replaces a regex match (or a subgroup thereof) with
+// a given string. The replacement string can use subgroup references like $1 or
+// ${groupname}, and can also use Go template expressions. An example is:
+//
+//   - regex: '([a-zA-Z]+), (?P<adjective>[a-zA-Z]+) world!'
+//     subgroup: 1
+//     with: 'fresh'
+//
+// This would transform a file containing "Hello, cool world!" to "Hello, fresh
+// world!".
+func actionRegexReplace(ctx context.Context, rr *model.RegexReplace, sp *stepParams) error {
+	uncompiled := make([]model.String, len(rr.Replacements))
+	for i, rp := range rr.Replacements {
+		uncompiled[i] = rp.Regex
+	}
+	compiledRegexes, err := templateAndCompileRegexes(uncompiled, sp.inputs)
+	if err != nil {
+		return err
+	}
+
+	for i, rp := range rr.Replacements {
+		subexps := compiledRegexes[i].NumSubexp()
+		if max := maxSubgroup([]byte(rp.With.Val)); max > subexps {
+			// Note to maintainers: subgroups are 1-indexed, and index i
+			// corresponds to subgroup i because subgroup 0 is just the whole
+			// regex match. Therefore we use ">" in the check above, and
+			// NumSubexp() is the index of the final subgroup.
+			return model.ErrWithPos(rp.With.Pos, "subgroup $%d is out of range; the largest subgroup in this regex is %d", max, subexps) //nolint:wrapcheck
+		}
+	}
+
+	for _, p := range rr.Paths {
+		if err := walkAndModify(p.Pos, sp.fs, sp.scratchDir, p.Val, func(b []byte) ([]byte, error) {
+			for i, rr := range rr.Replacements {
+				cr := compiledRegexes[i]
+
+				// Why reverse()? We have to replace starting at the end of the
+				// file working toward the beginning, so when we replace part of
+				// the buffer it doesn't invalidate the indices of the other
+				// matches indices.
+				allMatches := reverse(cr.FindAllSubmatchIndex(b, -1))
+
+				var err error
+				b, err = replaceWithTemplate(allMatches, b, rr, cr, sp.inputs)
+				if err != nil {
+					return nil, err
+				}
+			}
+			return b, nil
```

Comment:
I'm not sure I understand why we're returning the last replacement?

Reply:
The result of each transformation is pipelined into the following transformation, and the final result is returned. There's nothing special about the last replacement. Note the `=` rather than `:=` in `b, err = replaceWithTemplate(...`.

Reply:
Ah I missed the non-assignment. Thanks.